### PR TITLE
Expose `Check` command API

### PR DESCRIPTION
**Pull request by maximedenes**
This is useful for VsCoq2, in order to query Coq without hacking through the parser nor duplicating code.

### CODE REVIEW

File: vernac/comCheck.mli [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+val check_may_eval :
+  Evd.evar_map ->
+  Environ.env ->
+  (Environ.env -> Evd.evar_map -> Genredexpr.raw_red_expr -> Evd.evar_map * Redexpr.red_expr) ->
```
Comment (SkySkimmer) [inline vernac/comCheck.mli:14]: I don't like having this in the API
Maybe we can put the hook in redexpr.ml instead of vernacentries?

File: vernac/comCheck.mli [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  Genredexpr.raw_red_expr option ->
+  Constrexpr.constr_expr -> Pp.t
```
Comment (SkySkimmer) [inline vernac/comCheck.mli:16]:
```suggestion
  Constrexpr.constr_expr ->
  Pp.t
```

**Review by SkySkimmer — COMMENTED** [review]
Maybe it would make sense to have `Type` (`vernac_global_check`) in the same file?

**Review by ejgallego — APPROVED** [review]
LGTM, couple of suggestions:
- as Gaëtan says give a name to the `f_interp` type
- it is better to return a record with the result components and perform the final Pp.t composition in the client, for example for layout or translation purposes.

File: tactics/redexpr.mli [hunk 3]
```diff
@@ -22,6 +22,14 @@ type red_expr =
 
 type red_expr_val
 
+type interp_redexp = Environ.env -> Evd.evar_map -> Genredexpr.raw_red_expr -> Evd.evar_map * red_expr
+
+val f_interp_redexp : interp_redexp Hook.value
```
Comment (SkySkimmer) [inline tactics/redexpr.mli:27]:
```suggestion
val interp_redexp : interp_redexp
```
and eta expand in the .ml

File: tactics/redexpr.mli [hunk 4]
```diff
@@ -22,6 +22,14 @@ type red_expr =
...
+
+(** Hook to disappear when #8240 is fixed *)
```
Comment (SkySkimmer) [inline tactics/redexpr.mli:29]:
```suggestion
(** Hook to disappear when #12948 is fixed *)
```
(it got closed and reopened under a new number)

File: vernac/comCheck.ml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  env : Environ.env;
+  sigma : Evd.evar_map;
+  judgment : EConstr.unsafe_judgment;
+  redfun : Reductionops.reduction_function;
+  rc : Constrexpr.constr_expr;
+}
+
+type global_check_result = {
+  env : Environ.env;
+  sigma : Evd.evar_map;
```
Comment (SkySkimmer) [inline vernac/comCheck.ml:28]: really weird that this has an evar map

File: vernac/comCheck.mli [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  Evd.evar_map ->
+  Constrexpr.constr_expr ->
+  check_result * Univ.ContextSet.t
+
+val eval :
+  Environ.env ->
+  Evd.evar_map ->
+  Genredexpr.raw_red_expr ->
+  Constrexpr.constr_expr ->
+  eval_result * Univ.ContextSet.t
```
Comment (ejgallego) [inline vernac/comCheck.mli:43]: Looks very nice! A couple of further wishlists if you feel like doing an iteration:
- wrapping the result in a record `'a With_contextSet.t` would help having nicer serialization
- Maybe some ocamldocs could be helpful to the reader.

File: vernac/comCheck.ml (hunk 5 above)
Comment (ejgallego) [reply vernac/comCheck.ml:28]: That's for the printing, the result may be open so you need to interpret evars in printing IIUC.

I was thinking of a simpler result type, but now that I see what Maxime has done it looks better this way to me. Ideally we would have something like a "printable object" record.

Comment (SkySkimmer) [reply vernac/comCheck.ml:28]: >That's for the printing, the result may be open so you need to interpret evars in printing IIUC.

The whole point of global_check is that the result is not open.

Comment (maximedenes) [reply vernac/comCheck.ml:28]: The code may be simplified if we have some invariants about terms being evar-free. Note that some operations on universes are also performed in the evar map.

Anyway, IMHO such improvements are orthogonal to this PR, which only exposes the current API (hopefully in a way that is explicit and robust enough). I just did a tiny bit of cleanup to remove the dummy half-factorization of `Eval` and `Check` which was very strange style and was making the types much less precise.

Comment (SkySkimmer) [reply vernac/comCheck.ml:28]: >the dummy half-factorization of Eval and Check which was very strange style and was making the types much less precise.

Shouldn't it have been factorized more instead? AFAICT we return the reduction function and the judgment to call print_eval on them, but why don't we instead return the reduced judgment?

To the user the difference between `Eval red in term` and `Check ltac:(let c := open_constr:(term) in let c := eval red in c in refine c)` is approximately that Check prints info about evars and Eval doesn't, and maybe some subtleties about ltac that don't really matter since I only used ltac because there's no other way to Eval a subterm.

In other words `Check term` should be equivalent to `Eval identity in term`.

So IMO we should factorize more (and remove the differences we don't care about) instead of less.

>Anyway, IMHO such improvements are orthogonal to this PR, which only exposes the current API (hopefully in a way that is explicit and robust enough).

This is untrue, this PR clearly modifies the API. If you just want to expose the API then do that and splitting printing and compute can be done in a followup PR.

Comment (maximedenes) [reply vernac/comCheck.ml:28]: Ok, I don't care about that part so I'll open a more uncontroversial PR.

**Comment by maximedenes** [comment]
Ok, I reduced the content of the PR to what we'd like to have for VsCoq2 (straightforward exposition of the API), as the other changes will take too much effort.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Comment by maximedenes** [comment]
Thanks!
